### DISCUSSION TITLE
allow defining a method for fixed arrays of any length

### DISCUSSION THREAD

**Discussion by JalonSolov** [vlang/v#11871] [Ideas]
(From a discussion on discord...)

The problem right now is that you cannot use the same methods on fixed arrays as you can on dynamic arrays.  For example:
```v
mut buf = [10]byte{}
buf[0] = 0x61
println(buf.bytestr())
```
give the error
```
bug.v:3:13: error: unknown method: `[10]byte.bytestr`
    1 | mut buf := [10]byte{}
    2 | buf[0] = 0x61
    3 | println(buf.bytestr())
      |             ~~~~~~~~~
    4 |
```
Meaning there's no way to print the contents of the array as a string without converting to a dynamic array, first.

Suggested syntax would be something like
```v
fn (x [_]byte) foo()
```
or
```v
fn (x [*]byte) foo()
```

**Comment by dumblob** [comment]
Why new syntax? Sounds too ad-hoc to me.

If methods for stdlib/built-in modules, then [full-featured](https://github.com/vlang/v/issues/2025 ) (see also https://github.com/vlang/v/commit/895cefe351381955bd38057f581c9bc7a20598a3 and https://github.com/vlang/v/issues/9081#issuecomment-832507515 and https://github.com/vlang/v/discussions/7610#discussioncomment-750216 and https://github.com/vlang/v/discussions/7610#discussioncomment-691291 ).

But still I'm surprised the "dynamic array methods" can't be specialized by "compiler magic" to fixed arrays. Sure, there'll be some code duplication to make fixed-array methods faster than dynamic array methods, but I don't think it'll be too much code (maybe 100-200 SLOC of "unshared" similar code). And if speed is of no concern, then I'd just wrap existing dynamic array methods to make them compatible with fixed size arrays out of the box.

But new syntax - that's an overkill IMHO.

**Comment by JalonSolov** [comment]
Using the `_` isn't new syntax, the only thing "new" about it would be allowing it in function signatures as shown.

**Comment by dumblob** [comment]
Well, it assigns a new meaning to the "existing" syntax. How would it read then?

Admittedly I really dislike making yet another exception for some *V* built-in constraint/deficiency. Either full support for "methods for foreign modules" or nothing.

I'd emphasize though, that in this case I don't see a reason for either of these "solutions". I'm certain this is much better addressed by compiler magic rather than extending *V* **language** (i.e. the syntax & semantics).

**Comment by JalonSolov** [comment]
This has nothing to do with methods for foreign modules.  It is to avoid having to create a separate function for every possible value between the `[` and `]` characters, similar to this idea:
```v
fn (x [<T>]byte) foo()
```
... but the problem with that is that V currently doesn't allow generics in builtins.

**Comment by dumblob** [comment]
Ok, thanks for explanation. Then I'm curious what you all will choose in the end - keep me posted :wink:. Mind that if the `[_]` syntax will win, then it might get difficult in the future to use generics with bultins (and I think it's quite probable builtins will at some point after *V 1.0* use generics to some degree).
